### ISSUE TITLE
can some miner threads be paused while the other are running without affecting the hash results

### ISSUE THREAD

**Issue by cppdev-123** [fireice-uk/xmr-stak#2275]
I already forked xmr-stak and going to add some features to it before pulling
I managed to add pause, resume, stop to the miner
also the miner can now change the command line on the fly
but I want to make the miner compatible with the idle status of the system
e.g use no more than 50% when system isn't idle , and use as much as possible when the system is idle
I implemented this by restarting the miner threads (stop executor , then start it again in idle mode enabled)

but I look for another solution : the threads that exceeds the certain idle threads count is paused by default and only allowed to resume when the system is idle

this is easy to implement but I don't know whether this will result in rejected shares or not ?
I am not expert on the hashing algorithm itself .

**Comment by Spudz76** [comment]
Jobs are sort of like raffle tickets, they just need to be stapled together in the correct size (difficulty / pool job size) before sending upstream.  However torn or illegible tickets are invalid - so you can't staple an incomplete run of hashes (torn) nor failed recheck (the CPU miner doubles as the validator, look at what parts are still enabled when `-DCPU_ENABLE=NO`).  So you can pause wherever a work-run has completed and the result has been accepted for absolute sure, it is the best time.  Other times have varying loss of work (from aborting before a result that can be stapled) or from being idle between when you've unpaused it and when it can find its next work beginning (could be 100ms, could be 15s).

The smaller the difficulty the more often it hits that pause point, but at expense of network traffic, and difficulty (work batch size) comes from the pool you connect to.

The miners take jobs from the executor much like a very small pool (shared mem instead of a network though), so whenever a miner thread is in between jobs, you can pause it (or cork its incoming job flow, which causes it to go idle such as it would if your Internet unplugged).  The executor shouldn't care as long as it awaits the current job to hit some form of completion (next bite of cake).  The downside to pausing it is also that the jobs are in chunks so you have to wait until the next job is available once you've uncorked (re-subscribed).

So I guess the answer is YES you can pause it whenever and NO it won't always crash results, but there are also plenty of places where it would.  I am not specifically sure of all the executor pathways or how they work though.

My main question is how you can tell if the system is busy clogged with non-mining work or just busy mining, since both are CPU usage, and there are no good ways to subtract specific processes from that.  Once you're using 100% of all cores mining, how would even know automatically if some other task wanted to use CPU (without leaving at least one core unused as a "canary", or dropping a thread to take a peek now and then).

**Comment by cppdev-123** [comment]
Idle status is entered when two conditions are meet :
1 - system cpu usage is low

I measure the cpu usage of the system with NtQuerySystemInformation Which probably is the most accurate method and is used also by task manager 
Then measure the miner usage with GetProcessTimes 

Then sub the miner usage from the system usage

The idle threads have low priority so when  the system needs to use the cpu idle threads will give most their cpu slice to the system

2 - when no input is entered for particular amount of time e.g 1 minute

From what you said the best way to prevent idle threads from running when they mustn't run is to pause on the point they check for the new job
And if they are already working their job may be cancelled and return to check for a job

**Comment by Spudz76** [comment]
yes one way you await the finish and then not provide a following job

But I have had GPUs fail and hang themselves while the CPU threads kept mining just fine (at the reduced rate of course) so I know you can also be rough with them (but maybe lose some hashwork that can't be ticketed)

**Comment by cppdev-123** [comment]
The threads are already checking the job number many often like this : 
while (globalStates::inst().iGlobalJobNo.load(std::memory_order_relaxed) == iJobNo && !bQuit)

So it's easy to add checking for idle for particular threads
